tools/checklocks/state.go

```diff
@@ -184,7 +184,46 @@ func (l *lockState) count() int {
 
 // isCompatible returns true if the states are compatible.
 func (l *lockState) isCompatible(other *lockState) bool {
-	return l.isSubset(other) && other.isSubset(l)
+	if l == nil || other == nil {
+		return true
+	} else {
+		return l.isSubset(other) && other.isSubset(l)
+	}
+}
+
+// join returns a new lockState of the intersection of 2 states.
+// nil is consider the superset of locks, holding all locks in the other set.
+func (l *lockState) join(other *lockState) *lockState {
+	if l == nil {
+		return other
+	} else if other == nil {
+		return l
+	}
+
+	rls := l.fork()
+
+	for addr, val := range rls.stored {
```

Comment:
This is a bit of a complex operation. I would recommend a comment explaining the updates to stored.
``` 
 // Take the intersection of stored elements.
  for addr, val ...
```
Similar at the very top of `present :=` to explain lockedMutexes.